modules/mod-deep-learning/ModelCard.cpp

```diff
@@ -0,0 +1,371 @@
+      if(!doc[key.c_str()].IsUint64())
+         throwTypeError(key, "uint64", doc);
+      
+      return doc[key.c_str()].GetUint64();
+   }
+
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc, uint64_t defaultValue)
+   {
+      try
+      {
+         return tryGetUint64(key, doc);
+      }
+      catch (const InvalidModelCardDocument &e)
+      {
+         return defaultValue;
+      }
+   }
+}
+
+namespace parsers 
+{
+   Doc ParseString(const std::string &data)
+   {
+      Doc d;
+      // parse the data
+      d.Parse(data.c_str());
+      if (d.Parse(data.c_str()).HasParseError()) 
+      {
+         TranslatableString msg = XO("Error parsing JSON from string:\n%s\nDocument: %s ")
+                                       .Format(wxString(rapidjson::GetParseError_En(d.GetParseError())), 
+                                                wxString(data));
+         throw InvalidModelCardDocument(msg, "");
+      }
+      
+      return d;
+   }
+
+   Doc ParseFile(const std::string &path)
+   {
+      wxString docStr;
+      wxFile file = wxFile(path);
+
+      if(!file.ReadAll(&docStr))
+         throw InvalidModelCardDocument(XO("Could not read file."), "");
+
+      return ParseString(audacity::ToUTF8(docStr));
+   }
+}
+
+// ModelCard Implementation
+
+void ModelCard::Validate(const Doc& doc, const Doc& schema)
+{
+   rapidjson::SchemaDocument schemaDoc(schema);
+   rapidjson::SchemaValidator validator(schemaDoc);
+
+   if (!doc.Accept(validator)) 
+   {
+      // Input JSON is invalid according to the schema
+      // Output diagnostic information
+      TranslatableString msg = XO(
+         "A schema violation was found in the Model Card.\n "
+         "- violation found in URI: %s\n"
+         "- schema field violated: %s\n"
+         "- invalid document: %s\n"
+         "- schema document: %s\n"
+      );
+
+      rapidjson::StringBuffer sb;
+
+      validator.GetInvalidSchemaPointer().StringifyUriFragment(sb);
+      wxString uri(sb.GetString());
+      sb.Clear();
+
+      wxString invalidSchemaField(validator.GetInvalidSchemaKeyword());
+
+      rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+      doc.Accept(writer);
+      wxString invalidDocument(sb.GetString());
+      sb.Clear();
+   
+      rapidjson::Writer<rapidjson::StringBuffer> swriter(sb);
+      schema.Accept(swriter);
+      wxString schemaDocument(sb.GetString());
+
+      throw InvalidModelCardDocument(msg.Format(uri, invalidSchemaField, 
+                                                invalidDocument, schemaDocument), "");
+   }
+}
+
+void ModelCard::SerializeToFile(const std::string &path) const
+{
+   rapidjson::StringBuffer sb;
+   rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+
+   writer.StartObject();
+   Serialize(writer);
+   writer.EndObject();
+
+   wxFile file(path, wxFile::write);
+   if (!file.Write(wxString(sb.GetString())))
+      throw InvalidModelCardDocument(XO("Could not serialize ModelCard to file"), "");
```

Comment:
Perhaps the error message should quote the path to be more helpful to the user.

Perhaps call the static member function `TranslatableString FileException::WriteFailureMessage(const wxFileName &fileName);` to format the message consistently with other messages in Audacity.

Reply:
b447514d38b84c98c1d08508c2f6d12ca752d1c0